In [ ]:
# Instala o PuLP
# https://pypi.org/project/PuLP/
# https://coin-or.github.io/pulp/
!pip install -q pulp

In [1]:
# Imports
from pulp import *
import warnings
warnings.filterwarnings("ignore", category=FutureWarning)

### Parâmetros

- Ai = Número máximo de smartphones modelo tipo i para vender este mês, onde i pertence ao conjunto {Xiaomi, Iphone}
- Bi = Preço de venda de smartphones modelo tipo i, onde i pertence ao conjunto {Xiaomi, Iphone}
- Ci = Preço de custo das peças componentes para smartphones modelo tipo i, onde i pertence ao conjunto {Xiaomi, Iphone}
- Di = Custo de mão de obra de montagem por hora de smartphones modelo tipo i, onde i pertence ao conjunto {Xiaomi, Iphone}
- Ei = Custo de mão de obra de teste por hora de smartphones modelo tipo i, onde i pertence ao conjunto {Xiaomi, Iphone}
- F = Número máximo de horas de trabalho de montagem
- G = Número máximo de horas de trabalho de teste
- Hf,i = Horas de montagem necessárias para construir cada modelo de smartphone tipo i, onde i pertence ao conjunto {Xiaomi, Iphone}
- Hg,i = Horas de teste necessárias para testar cada modelo de smartphone tipo i, onde i pertence ao conjunto {Xiaomi, Iphone}

### Variável de Decisão

- Xi = Número de smartphones modelo tipo i a produzir este mês, onde i pertence ao conjunto {Xiaomi, Iphone}

### Função Objetivo

![title](imagens/formula.png)

### Restrições

- O número de smartphones modelo tipo i a serem produzidos não pode ser negativo, ou seja, Xi >= 0, onde i pertence ao conjunto {Xiaomi, Iphone}.


- O número total de horas de montagem não pode ser maior que o número máximo de horas de mão de obra de montagem disponíveis.


- O número total de horas de teste não pode ser maior do que o máximo de horas de mão de obra de teste disponíveis.


- O número de smartphones modelo tipo i a serem produzidos não pode ser maior do que o número máximo de smartphones modelo tipo i a serem vendidos neste mês, onde i pertence ao conjunto {Xiaomi, Iphone}.

### Organizando os Parâmetros do modelo matemático

In [3]:
# Número máximo de smartphones para vender este mês
A_Xiaomi = 600
A_Iphone = 1200

In [4]:
# Preço de venda de smartphones 
B_Xiaomi = 300 
B_Iphone = 450 

In [5]:
# Preço de custo das peças componentes para smartphones
C_Xiaomi = 150
C_Iphone = 225

In [6]:
# Custo de mão de obra de montagem por hora de smartphones
D_Xiaomi = 11
D_Iphone = 11

In [7]:
# Custo de mão de obra de teste por hora de smartphones 
E_Xiaomi = 15 
E_Iphone = 15 

In [8]:
# Número máximo de horas de trabalho de montagem
F = 10000 

In [9]:
# Número máximo de horas de trabalho de teste
G = 3000 

In [10]:
# Horas de montagem necessárias para construir cada modelo de smartphone
Hfi_Xiaomi = 5
Hfi_Iphone = 6

In [11]:
# Horas de teste necessárias para testar cada modelo de smartphone
Hgi_Xiaomi = 1 
Hgi_Iphone = 2

### Criando var para otimizar

In [12]:
problema = LpProblem("MixProdutos", LpMaximize)
problema

MixProdutos:
MAXIMIZE
None
VARIABLES

### Definindo var de tomada de decisão para cada modelo de smartphone

In [13]:
# Define as variáveis
x_Xiaomi = LpVariable("Unidades Lua1", 0, None, LpInteger)
x_Iphone = LpVariable("Unidades Lua2", 0, None, LpInteger)

In [14]:
# Print
print(x_Xiaomi)
print(x_Iphone)

Unidades_Lua1
Unidades_Lua2


### Função Lucro total

![title](imagens/formula.png)

In [15]:
# Faturamento
faturamento = (x_Xiaomi * B_Xiaomi) + (x_Iphone * B_Iphone)
faturamento

300*Unidades_Lua1 + 450*Unidades_Lua2 + 0

In [16]:
# Custo de Montagem
custo_montagem = (x_Xiaomi * Hfi_Xiaomi * D_Xiaomi) + (x_Iphone * Hfi_Iphone * D_Iphone)
custo_montagem

55*Unidades_Lua1 + 66*Unidades_Lua2 + 0

In [17]:
# Custo de Teste
custo_teste = (x_Xiaomi * Hgi_Xiaomi * E_Xiaomi) + (x_Iphone * Hgi_Iphone * E_Iphone)
custo_teste

15*Unidades_Lua1 + 30*Unidades_Lua2 + 0

In [18]:
# Custo de Componentes
custo_componentes = (x_Xiaomi * C_Xiaomi) + (x_Iphone * C_Iphone) 
custo_componentes

150*Unidades_Lua1 + 225*Unidades_Lua2 + 0

In [19]:
# Lucro = Faturamento - Custo de Montagem - Custo de Teste - Custo de Componentes
problema += faturamento - custo_montagem - custo_teste - custo_componentes
problema

MixProdutos:
MAXIMIZE
80*Unidades_Lua1 + 129*Unidades_Lua2 + 0
VARIABLES
0 <= Unidades_Lua1 Integer
0 <= Unidades_Lua2 Integer

In [20]:
# Número máximo de horas de montagem
problema += (x_Xiaomi * Hfi_Xiaomi) + (x_Iphone * Hfi_Iphone) <= F,"Número Máximo de Horas de Montagem"

In [21]:
# Número máximo de horas de teste
problema += (x_Xiaomi * Hgi_Xiaomi) + (x_Iphone * Hgi_Iphone) <= G,"Número Máximo de Horas de Teste"

In [22]:
# Produção menor ou igual a demanda pelo modelo Lua1
problema += x_Xiaomi <= A_Xiaomi,"Produção menor ou igual a demanda pelo modelo Lua1"

In [23]:
# Produção menor ou igual a demanda pelo modelo Lua2
problema += x_Iphone <= A_Iphone,"Produção menor ou igual a demanda pelo modelo Lua2"

In [29]:
problema

MixProdutos:
MAXIMIZE
80*Unidades_Lua1 + 129*Unidades_Lua2 + 0
SUBJECT TO
Número_Máximo_de_Horas_de_Montagem: 5 Unidades_Lua1 + 6 Unidades_Lua2 <= 10000

Número_Máximo_de_Horas_de_Teste: Unidades_Lua1 + 2 Unidades_Lua2 <= 3000

Produção_menor_ou_igual_a_demanda_pelo_modelo_Lua1: Unidades_Lua1 <= 600

Produção_menor_ou_igual_a_demanda_pelo_modelo_Lua2: Unidades_Lua2 <= 1200

VARIABLES
0 <= Unidades_Lua1 Integer
0 <= Unidades_Lua2 Integer

### Usando Solve para resolver.

In [25]:
# Otimização
problema.solve()

1

In [26]:
# Lucro Maximizado
print("Lucro Maximizado:", value(problema.objective))

Lucro Maximizado: 199600.0


In [27]:
# Número de Unidades do Modelo Lua 1 a Produzir
print("Número de Unidades do Modelo Lua 1 a Produzir:", problema.variables()[0].varValue)

Número de Unidades do Modelo Lua 1 a Produzir: 560.0


In [28]:
# Número de Unidades do Modelo Lua 1 a Produzir
print("Número de Unidades do Modelo Lua 2 a Produzir:", problema.variables()[1].varValue)

Número de Unidades do Modelo Lua 2 a Produzir: 1200.0


A empresa X Tech deve produzir 560 unidades do tipo de smartphone Xiaomi e 1200 do tipo Iphone para atingir o lucro máximo de R$199.600.